File: packages/vaultkeeper/src/identity/manifest.ts
/**
 * Trust manifest management — load, save, and query approved executable hashes.
 */

import * as fs from 'node:fs/promises'
import * as path from 'node:path'
import { FilesystemError } from '../errors.js'
import type { TrustManifest, TrustManifestEntry } from './types.js'

const MANIFEST_FILENAME = 'trust-manifest.json'

/** Raw shape persisted to disk (plain object, not a Map). */
interface RawManifest {
  version: number
  entries: Record<string, TrustManifestEntry>
}

/**
 * Type guard that checks `value` is an object with a `version: number` and
 * `entries` object property, matching {@link RawManifest}.
 *
 * We avoid type assertions by narrowing through the `in` operator and
 * `typeof` checks. The discriminated union `unknown → object` chain lets
 * TypeScript track narrowness without any `as` casts.
 */
function isRawManifest(value: unknown): value is RawManifest {
  if (typeof value !== 'object' || value === null) return false
  if (!('version' in value) || typeof value.version !== 'number') return false
  if (!('entries' in value) || typeof value.entries !== 'object' || value.entries === null)
    return false
  return true
}

/**
 * Type guard for a single {@link TrustManifestEntry}.
 */
function isTrustManifestEntry(value: unknown): value is TrustManifestEntry {
  if (typeof value !== 'object' || value === null) return false
  if (!('hashes' in value) || !Array.isArray(value.hashes)) return false
  if (!('trustTier' in value)) return false
  const { trustTier } = value
  if (trustTier !== 1 && trustTier !== 2 && trustTier !== 3) return false
  return true
}

/**
 * Load the trust manifest from `configDir`.
 * Returns an empty `Map` if the manifest file does not yet exist.
 * @throws {@link FilesystemError} If the file exists but cannot be read, or its
 *   contents are not valid JSON.
 * @internal
 */
export async function loadManifest(configDir: string): Promise<TrustManifest> {
  const manifestPath = path.join(configDir, MANIFEST_FILENAME)
  let rawText: string
  try {
    rawText = await fs.readFile(manifestPath, 'utf8')
  } catch (err) {
    if (typeof err === 'object' && err !== null && 'code' in err && err.code === 'ENOENT') {
      return new Map()
    }
    const detail = err instanceof Error ? err.message : String(err)
    throw new FilesystemError(
      `Cannot read trust manifest at ${manifestPath}: ${detail}`,
      manifestPath,
      'read',
    )
  }

  let parsed: unknown
  try {
    parsed = JSON.parse(rawText)
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err)
    throw new FilesystemError(
      `Trust manifest at ${manifestPath} is not valid JSON: ${detail}`,
      manifestPath,
      'read',
    )
  }

  if (!isRawManifest(parsed)) {
    return new Map()
  }

  const manifest: TrustManifest = new Map()
  for (const [namespace, entry] of Object.entries(parsed.entries)) {
    if (isTrustManifestEntry(entry)) {
      manifest.set(namespace, { hashes: [...entry.hashes], trustTier: entry.trustTier })
    }
  }
  return manifest
}

/**
 * Persist `manifest` to `configDir`, creating the directory if necessary.
 * @throws {@link FilesystemError} If the directory cannot be created or the
 *   manifest file cannot be written.
 * @internal
 */
export async function saveManifest(configDir: string, manifest: TrustManifest): Promise<void> {
  const manifestPath = path.join(configDir, MANIFEST_FILENAME)
  try {
    await fs.mkdir(configDir, { recursive: true })
    const entries: Record<string, TrustManifestEntry> = {}
    for (const [namespace, entry] of manifest) {
      entries[namespace] = entry
    }
    const raw: RawManifest = { version: 1, entries }
    await fs.writeFile(manifestPath, JSON.stringify(raw, null, 2), 'utf8')
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err)
    throw new FilesystemError(
      `Cannot write trust manifest at ${manifestPath}: ${detail}`,
      manifestPath,
      'write',
    )
  }
}

/**
 * Return a new manifest that includes `hash` under `namespace`.
 * If the namespace does not yet exist it is created with tier 3 (Unverified).
 * The trust tier of an existing entry is not changed.
 * @internal
 */
export function addTrustedHash(
  manifest: TrustManifest,
  namespace: string,
  hash: string,
): TrustManifest {
  const next = new Map(manifest)
  const existing = next.get(namespace)
  if (existing === undefined) {
    next.set(namespace, { hashes: [hash], trustTier: 3 })
  } else if (!existing.hashes.includes(hash)) {
    next.set(namespace, { hashes: [...existing.hashes, hash], trustTier: existing.trustTier })
  }
  return next
}

/**
 * Return `true` if `hash` is in the approved list for `namespace`.
 * @internal
 */
export function isTrusted(manifest: TrustManifest, namespace: string, hash: string): boolean {
  const entry = manifest.get(namespace)
  if (entry === undefined) return false
  return entry.hashes.includes(hash)
}

File: crates/vaultkeeper-core/src/identity/manifest.rs
//! Trust manifest management — load, save, and query approved executable hashes.

use super::types::{TrustManifest, TrustManifestEntry};
use crate::backend::HostPlatform;
use crate::errors::VaultError;
use crate::types::TrustTier;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

const MANIFEST_FILENAME: &str = "trust-manifest.json";

/// On-disk representation of the trust manifest.
#[derive(Debug, Serialize, Deserialize)]
struct RawManifest {
    version: u32,
    entries: HashMap<String, RawEntry>,
}

/// On-disk representation of a single manifest entry.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RawEntry {
    hashes: Vec<String>,
    trust_tier: u8,
}

/// Load the trust manifest from the config directory.
/// Returns an empty manifest if the file does not exist.
pub async fn load_manifest(host: &dyn HostPlatform) -> Result<TrustManifest, VaultError> {
    let manifest_path = host.config_dir().join(MANIFEST_FILENAME);

    match host.file_exists(&manifest_path).await {
        Ok(true) => {}
        _ => return Ok(HashMap::new()),
    }

    let content = host.read_file(&manifest_path).await?;
    let json = String::from_utf8(content)
        .map_err(|e| VaultError::Other(format!("Invalid UTF-8 in trust manifest: {e}")))?;

    let raw: RawManifest = serde_json::from_str(&json)
        .map_err(|e| VaultError::Other(format!("Failed to parse trust manifest: {e}")))?;

    let mut manifest = HashMap::new();
    for (namespace, entry) in raw.entries {
        let tier = match entry.trust_tier {
            1 => TrustTier::Sigstore,
            2 => TrustTier::Tofu,
            _ => TrustTier::Dev,
        };
        manifest.insert(
            namespace,
            TrustManifestEntry {
                hashes: entry.hashes,
                trust_tier: tier,
            },
        );
    }

    Ok(manifest)
}

/// Save the trust manifest to the config directory.
pub async fn save_manifest(
    host: &dyn HostPlatform,
    manifest: &TrustManifest,
) -> Result<(), VaultError> {
    let mut entries = HashMap::new();
    for (namespace, entry) in manifest {
        entries.insert(
            namespace.clone(),
            RawEntry {
                hashes: entry.hashes.clone(),
                trust_tier: entry.trust_tier as u8,
            },
        );
    }

    let raw = RawManifest {
        version: 1,
        entries,
    };

    let json = serde_json::to_string_pretty(&raw)
        .map_err(|e| VaultError::Other(format!("Failed to serialize trust manifest: {e}")))?;

    let manifest_path = host.config_dir().join(MANIFEST_FILENAME);
    host.write_file(&manifest_path, json.as_bytes(), 0o600)
        .await
}

/// Return a new manifest with `hash` added under `namespace`.
/// If the namespace does not exist, it is created with tier 3 (Dev/Unverified).
/// The trust tier of an existing entry is not changed.
pub fn add_trusted_hash(manifest: &TrustManifest, namespace: &str, hash: &str) -> TrustManifest {
    let mut next = manifest.clone();
    let entry = next
        .entry(namespace.to_string())
        .or_insert_with(|| TrustManifestEntry {
            hashes: Vec::new(),
            trust_tier: TrustTier::Dev,
        });
    if !entry.hashes.iter().any(|h| h == hash) {
        entry.hashes.push(hash.to_string());
    }
    next
}

/// Return `true` if `hash` is in the approved list for `namespace`.
pub fn is_trusted(manifest: &TrustManifest, namespace: &str, hash: &str) -> bool {
    manifest
        .get(namespace)
        .is_some_and(|entry| entry.hashes.iter().any(|h| h == hash))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn empty_manifest() -> TrustManifest {
        HashMap::new()
    }

    #[test]
    fn is_trusted_returns_false_for_empty_manifest() {
        assert!(!is_trusted(&empty_manifest(), "app", "abc123"));
    }

    #[test]
    fn add_trusted_hash_creates_new_namespace() {
        let m = add_trusted_hash(&empty_manifest(), "app", "hash1");
        assert!(is_trusted(&m, "app", "hash1"));
        assert!(!is_trusted(&m, "app", "hash2"));
    }

    #[test]
    fn add_trusted_hash_appends_to_existing() {
        let m = add_trusted_hash(&empty_manifest(), "app", "hash1");
        let m = add_trusted_hash(&m, "app", "hash2");
        assert!(is_trusted(&m, "app", "hash1"));
        assert!(is_trusted(&m, "app", "hash2"));
    }

    #[test]
    fn add_trusted_hash_deduplicates() {
        let m = add_trusted_hash(&empty_manifest(), "app", "hash1");
        let m = add_trusted_hash(&m, "app", "hash1");
        assert_eq!(m["app"].hashes.len(), 1);
    }

    #[test]
    fn add_trusted_hash_preserves_existing_tier() {
        let mut m = empty_manifest();
        m.insert(
            "app".to_string(),
            TrustManifestEntry {
                hashes: vec!["old".to_string()],
                trust_tier: TrustTier::Tofu,
            },
        );
        let m = add_trusted_hash(&m, "app", "new");
        assert_eq!(m["app"].trust_tier, TrustTier::Tofu);
    }

    #[test]
    fn new_namespace_gets_dev_tier() {
        let m = add_trusted_hash(&empty_manifest(), "new-app", "abc");
        assert_eq!(m["new-app"].trust_tier, TrustTier::Dev);
    }
}
